package/com.unity.formats.usd/Runtime/Scripts/IO/Materials/ShaderExporterBase.cs

```diff
@@ -33,33 +33,123 @@ public class ShaderExporterBase {
-
-      System.IO.File.Copy(srcPath, filePath, overwrite: true);
-
-      // Make file path baked into USD relative to scene file and use forward slashes.
-      filePath = ImporterBase.MakeRelativePath(scene.FilePath, filePath);
-      filePath = filePath.Replace("\\", "/");
-
-      var uvReader = new PrimvarReaderSample<Vector2>();
-      uvReader.varname.defaultValue = new TfToken("st");
-      scene.Write(usdShaderPath + "/uvReader", uvReader);
-      var tex = new TextureReaderSample(filePath, usdShaderPath + "/uvReader.outputs:result");
-      scene.Write(usdShaderPath + "/" + textureName, tex);
-      return usdShaderPath + "/" + textureName + ".outputs:" + textureOutput;
-#else
-      // Not supported at run-time, too many things can go wrong
-      // (can't encode compressed textures, etc).
-      throw new System.Exception("Not supported at run-time");
-#endif
-    }
+                                 string textureOutput)
+    {
+      // We have to handle multiple cases here:
+      // - file exists on disk
+      //   - file is a supported format => can be directly copied
+      //   - file is not in a supported format => need to blit / export
+      // - file is only in memory
+      //   - a Texture2D
+      //   - a Texture
+      //   - a RenderTexture
+      //   - needs special care if marked as Normal Map
+      //     (can probably only be detected in an Editor context, and heuristically at runtime)
+      //   => need to blit / export
+      // - file is not supported at all (or not yet)
+      //   - a 3D texture
+      //   => needs to be ignored, log Warning
+
+      bool textureIsExported = false;
+
+      string filePath = null;
+      string fileName = null;
+
+      var texture = material.GetTexture(textureName);
+
+  #if UNITY_EDITOR
+      var srcPath = UnityEditor.AssetDatabase.GetAssetPath(texture);
+      if(!string.IsNullOrEmpty(srcPath))
+      { 
+        srcPath = srcPath.Substring("Assets/".Length);
+        srcPath = Application.dataPath + "/" + srcPath;
+        fileName = System.IO.Path.GetFileName(srcPath);
+        filePath = System.IO.Path.Combine(destTexturePath, fileName);
+
+        if(System.IO.File.Exists(srcPath))
+        { 
+          // USDZ officially only supports png / jpg / jpeg
+          // https://graphics.pixar.com/usd/docs/Usdz-File-Format-Specification.html
 
+          var ext = System.IO.Path.GetExtension(srcPath).ToLowerInvariant();
+          if(ext == ".png" || ext == ".jpg" || ext == ".jpeg") { 
+            System.IO.File.Copy(srcPath, filePath, overwrite: true);
+            if (System.IO.File.Exists(filePath))
+              textureIsExported = true;
+          }
+        }
+      }
+  #endif
+      if (!textureIsExported)
+      {
+        fileName = texture.name + "_" + Random.Range(10000000, 99999999).ToString();
+  #if UNITY_EDITOR
+        if (UnityEditor.AssetDatabase.Contains(texture))
+          fileName = texture.name + "_png";
```

Comment:
Oh, I just noticed -- need braces here

Also, maybe add a comment to explain that this is just to avoid file collisions in the event that there already exists a file to export by the same name.

Reply:
Added the comment. Also, I removed the #if UNITY_EDITOR part because now that I think about it this will cause more trouble than help - textures with identical names in the AssetDB would override / collide, so I removed it again for now. Ideas for future improvement are in the TODO section in the comment.